src/basic/socket-util.h

```diff
@@ -147,6 +148,16 @@ int flush_accept(int fd);
 
 struct cmsghdr* cmsg_find(struct msghdr *mh, int level, int type, socklen_t length);
 
+#define SOCKADDR_LL_LEN(sa)                                             \
+        ({                                                              \
+                const struct sockaddr_ll *_sa = &(sa);                  \
+                unsigned long _mac_len = sizeof(_sa->sll_addr);         \
+                assert(_sa->sll_family == AF_PACKET);                   \
+                if (be16toh(_sa->sll_hatype) == ARPHRD_INFINIBAND)      \
+                        _mac_len = INFINIBAND_ALEN;                     \
+                offsetof(struct sockaddr_ll, sll_addr) + _mac_len;      \
+        })
```

Comment:
hmm, could you please also add a quick comment to this macro explaining briefly what the pitfall around sockaddr_ll is? if you come by this piece of code it isn't obvious why we do all this, and hence a comment would is definitely of use.